### PR TITLE
[pybind11] Fix debug link failure

### LINKED ISSUE #30642: [pybind11:x64-windows] debug link failure
**Host Environment**
- OS:  Windows10
- Compiler:   Microsoft Visual Studio Community 2022 (64-bit) - Current,   Version 17.5.3

**To Reproduce**
CMakeLists.txt file vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
cmake_minimum_required (VERSION 3.16)
project (firestoreTest)

set (CMAKE_CXX_STANDARD 11)
set (CMAKE_CXX_STANDARD_REQUIRED True)
set (CMAKE_TOOLCHAIN_FILE c:/projects/github/microsoft/vcpkg/scripts/buildsystems/vcpkg.cmake)

find_package (pybind11 REQUIRED)

add_executable (firestoreTest
	"firestoreTest.cpp" "firestoreTest.h"
	"pythonTest.cpp" "pythonTest.h"
	"main.cpp"
)

target_link_libraries (firestoreTest
	PRIVATE pybind11::embed
)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Steps to reproduce the behavior**
vcpkg install pybind11:x64-windows
Build with above CMakeLists.txt.
Reported error:   vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
>------ Build All started: Project: firestoreTest, Configuration: x64-Debug ------
  [1/1] Linking CXX executable firestoreTest.exe
  FAILED: firestoreTest.exe 
  cmd.exe /C "cd . && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=CMakeFiles\firestoreTest.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\10.0.19041.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\10.0.19041.0\x64\mt.exe --manifests  -- "C:\PROGRA~1\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\bin\Hostx64\x64\link.exe" /nologo CMakeFiles\firestoreTest.dir\firestoreTest.cpp.obj CMakeFiles\firestoreTest.dir\pythonTest.cpp.obj CMakeFiles\firestoreTest.dir\main.cpp.obj  /out:firestoreTest.exe /implib:firestoreTest.lib /pdb:firestoreTest.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console  C:\Projects\github\microsoft\vcpkg\installed\x64-windows\debug\lib\python310_d.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cmd.exe /C "cd /D "C:\Projects\trials\28.2 - Firestore test\firestoreTest\out\build\x64-debug" && C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -noprofile -executionpolicy Bypass -file C:/Projects/github/microsoft/vcpkg/scripts/buildsystems/msbuild/applocal.ps1 -targetBinary "C:/Projects/trials/28.2 - Firestore test/firestoreTest/out/build/x64-debug/firestoreTest.exe" -installedDir C:/Projects/github/microsoft/vcpkg/installed/x64-windows/debug/bin -OutVariable out""
  LINK Pass 1: command "C:\PROGRA~1\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\bin\Hostx64\x64\link.exe /nologo CMakeFiles\firestoreTest.dir\firestoreTest.cpp.obj CMakeFiles\firestoreTest.dir\pythonTest.cpp.obj CMakeFiles\firestoreTest.dir\main.cpp.obj /out:firestoreTest.exe /implib:firestoreTest.lib /pdb:firestoreTest.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console C:\Projects\github\microsoft\vcpkg\installed\x64-windows\debug\lib\python310_d.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\firestoreTest.dir/intermediate.manifest CMakeFiles\firestoreTest.dir/manifest.res" failed (exit code 1104) with the following output:
C:\Projects\trials\28.2 - Firestore test\firestoreTest\out\build\x64-Debug\LINK : fatal error LNK1104: cannot open file 'python310.lib'
  ninja: build stopped: subcommand failed.
Build All failed.
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- Apparently pybind11 finds dependent libraries on it's own, insisting on using the release version of python310.lib/dll.  But cmake insists on a particular use of debug/release directories.  Thus when the project is building a debug version, only python310_d.lib/dll will be managed and found.
- If I provide CMakeLists.txt the path to the python310.lib, it will link, but execution of the program will not find the python310.dll (because python310_d.dll has been copied to my executable directory).
- I'm not savy enough with ports to fix this.
- In my mind, pybind11 should not force the release version of python.    vcpkg install pybind11:x64-windows should work with a debug or release build.

### LINKED ISSUE #31311: [pybind11] messes with python3 autolink
seen in a few PRs trying to link `python3.lib` in **debug** builds.
This is due to pybind11 undefining `_DEBUG` on line 228 in `pybind11/detail/common.h` which messes how python3 autolinks. 
(Maybe we need to force `Py_DEBUG` somewhere?)

### PR DESCRIPTION
**Pull request by JonLiu1993**
Fixes #30642
Fixes #31311

Pybind11 finds dependent libraries on it's own, insisting on using the release version of `python310.lib/dll`. But cmake insists on a particular use of debug/release directories. Thus when the project is building a debug version, only `python310_d.lib/dll` will be managed and found.
````
>------ Build All started: Project: test-usage, Configuration: x64-Debug ------
  MSBuild version 17.5.1+f6fdcf537 for .NET Framework
  
    Checking Build System
    Building Custom Rule C:/Users/test/source/repos/test-usage/test-usage/CMakeLists.txt
    python.cpp
C:\Users\test\source\repos\test-usage\out\build\x64-Debug\test-usage\LINK : fatal error LNK1104: cannot open file 'python310.lib' 

Build All failed.
````
Add `target_compile_definitions (pybind11_headers INTERFACE Py_DEBUG) `  to force pybind11 to use debug build of python.

Tested link successfully by `pybind11:x64-windows`:
````
>------ Build All started: Project: test-usage, Configuration: x64-Debug ------
  MSBuild version 17.5.1+f6fdcf537 for .NET Framework
  
    Checking Build System
    Building Custom Rule C:/Users/test/source/repos/test-usage/test-usage/CMakeLists.txt
    python.cpp
       Creating library C:/Users/test/source/repos/test-usage/out/build/x64-Debug/test-usage/Debug/test_pybind11_module.lib and object C:/Users/test/source/repos/test-usage/out/build/x64-Debug/test-usage/Debug/test_pybind11_module.exp
    test_pybind11_module.vcxproj -> C:\Users\test\source\repos\test-usage\out\build\x64-Debug\test-usage\Debug\test_pybind11_module
    Building Custom Rule C:/Users/test/source/repos/test-usage/CMakeLists.txt

Build All succeeded.
````

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] ~~SHA512s are updated for each updated download~~
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/pybind11/fix-debug-link.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 0d93203..031088a 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -196,6 +196,7 @@ if(NOT TARGET pybind11_headers)
+ 
+   target_compile_features(pybind11_headers INTERFACE cxx_inheriting_constructors cxx_user_literals
+                                                      cxx_right_angle_brackets)
++  target_compile_definitions (pybind11_headers INTERFACE Py_DEBUG)
```
Comment (Osyotr) [inline ports/pybind11/fix-debug-link.patch:9]: This should be `$<$<CONFIG:Debug>:Py_DEBUG>`.
Btw, can't this be fixed in `vcpkg-cmake-wrapper.cmake`?

Comment (Neumann-A) [reply ports/pybind11/fix-debug-link.patch:9]: I think this needs to be fixed in python3 doesn't it ?

Comment (JonLiu1993) [reply ports/pybind11/fix-debug-link.patch:9]: @Osyotr @Neumann-A , Thanks for your review, do you means add `target_compile_definitions ` to ptthon3 's `vcpkg-cmake-wrapper.cmake`?

Comment (Hoikas) [reply ports/pybind11/fix-debug-link.patch:9]: @Neumann-A pyconfig.h from python3 already has this:

```C
#ifdef _DEBUG
#       define Py_DEBUG
#endif
```

So no change is needed to the python3 port.

**Review by LilyWangLL — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for the fix @JonLiu1993 ! Thanks for checking it over @Osyotr @Neumann-A and @Hoikas !

File: ports/pybind11/fix-debug-link.patch (hunk 1 above)
Comment (zyrkiel) [reply ports/pybind11/fix-debug-link.patch]: Even if these lines are in pyconfig.h.in I don't find them in the generated pyconfig.h in my vcpkg buildtree...

Comment (JonLiu1993) [reply ports/pybind11/fix-debug-link.patch]: @zyrkiel, This PR has been merged. Please resubmit an issue and attach your error log. thanks.

**Pull request merged**